skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BtSenderActivity.java

```diff
@@ -148,18 +158,81 @@ private Disposable addUploadEventSubscription() {
     /**
      * Enable the bluetooth discovery for other devices. The timeout is specific seconds.
      */
-    private void enableDiscovery() {
+    @NeedsPermission({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void enableDiscovery() {
         Intent discoverableIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_DISCOVERABLE);
         // set the discovery timeout for 120s.
         discoverableIntent.putExtra(BluetoothAdapter.EXTRA_DISCOVERABLE_DURATION, CONNECT_TIMEOUT);
-        startActivity(discoverableIntent);
+        startActivityForResult(discoverableIntent, DISCOVERABLE_CODE);
+    }
+
+    @SuppressWarnings("CallNeedsPermission")
+    private void enableDiscoveryWithPermissionCheck() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            BtSenderActivityPermissionsDispatcher.enableDiscoveryWithPermissionCheck(this);
+        } else {
+            enableDiscovery();
+        }
+    }
+
+    /**
+     * Show the rationale dialog for location permission, only if the permission was granted,
+     * we can start using the bluetooth.
+     */
+    @OnShowRationale({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void showLocationPermissionDialog(final PermissionRequest request) {
+        new androidx.appcompat.app.AlertDialog.Builder(this)
+                .setCancelable(false)
+                .setMessage(R.string.permission_location_rationale)
+                .setPositiveButton(R.string.button_allow, (dialog, button) -> request.proceed())
+                .setNegativeButton(R.string.button_deny, (dialog, button) -> request.cancel())
+                .show();
+    }
+
+    /**
+     * If the permission was denied, finishing this activity.
+     */
+    @OnPermissionDenied({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void showDeniedForLocation() {
+        Toast.makeText(this, R.string.permission_location_never_ask, Toast.LENGTH_SHORT).show();
+        finish();
+    }
+
+    /**
+     * If clicked the "never ask", we should show a toast to guide user.
+     */
+    @OnNeverAskAgain({Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION})
+    void showNeverAskForCamera() {
+        Toast.makeText(this, R.string.permission_location_never_ask, Toast.LENGTH_SHORT).show();
```

Comment:
Rather than printing toast, it would be helpful for a user if we show some dialog where he can choose to navigate to settings directly. Otherwise, he may have to do a lot of efforts in searching the app in settings and then enable the location.